client/lib/client.dart

```diff
@@ -0,0 +1,717 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import 'dart:async';
+
+import 'package:http/http.dart' as http;
+import 'package:protobuf/protobuf.dart';
+import 'package:spine_client/firebase_client.dart';
+import 'package:spine_client/google/protobuf/field_mask.pb.dart';
+import 'package:spine_client/spine/base/error.pb.dart' as pbError;
+import 'package:spine_client/spine/base/field_path.pb.dart';
+import 'package:spine_client/spine/client/filters.pb.dart';
+import 'package:spine_client/spine/client/query.pb.dart';
+import 'package:spine_client/spine/client/subscription.pb.dart' as pbSubscription;
+import 'package:spine_client/spine/core/ack.pb.dart';
+import 'package:spine_client/spine/core/command.pb.dart';
+import 'package:spine_client/spine/core/diagnostics.pb.dart';
+import 'package:spine_client/spine/core/event.pb.dart';
+import 'package:spine_client/spine/core/tenant_id.pb.dart';
+import 'package:spine_client/spine/core/user_id.pb.dart';
+import 'package:spine_client/spine/time/time.pb.dart';
+import 'package:spine_client/spine/web/firebase/subscription/firebase_subscription.pb.dart';
+import 'package:spine_client/src/actor_request_factory.dart';
+import 'package:spine_client/src/any_packer.dart';
+import 'package:spine_client/src/http_client.dart';
+import 'package:spine_client/src/json.dart';
+import 'package:spine_client/src/known_types.dart';
+import 'package:spine_client/src/query_processor.dart';
+import 'package:spine_client/subscription.dart';
+import 'package:spine_client/validate.dart';
+
+/// A factory of [Client]s.
+///
+/// Creates [Clients] for posting commands, sending queries, and managing subscriptions on behalf
+/// of a certain user.
+///
+/// Example:
+/// ```dart
+/// import 'package:spine_client/client.dart';
+/// import './types.dart' as myTypes;
+///
+/// var clients = Clients(
+///     'https://example.org/',
+///     firebase: WebFirebaseClient(FirebaseApp().database),
+///     typeRegistries: [myTypes.types()]
+/// );
+/// var actor = UserId()
+///     ..value = 'some-user-id';
+/// var client = clients.onBehalfOf(actor);
+///
+/// // ...
+///
+/// var cmdRequest = client.command(postComment);
+/// var events = cmdRequest.observeEvents(CommentPosted());
```

Comment:
Do we have to instantiate the `CommentPosted` here?

I was browsing through the Dart Q&A @ StackOverflow and saw a couple of examples of passing the type as 
```
var events = cmdRequest.observeEvents(CommentPosted);
```
which looks more natural to me.

Just in case, here is [one of the questions answered](https://stackoverflow.com/questions/23094392/passing-a-class-type-as-a-variable-in-dart).

Reply:
We're going to use `cmdRequest.observeEvents<CommentPosted>()`.

Turns out that Dart, just like Kotlin, has reified generics, meaning that we can use the type argument as a regular type inside, say, a method.